netplugin/netd.go

```diff
@@ -14,248 +14,298 @@ limitations under the License.
+	app.Usage = "Contiv netplugin service"
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:   "log-level",
+			Value:  "INFO",
+			EnvVar: "CONTIV_NETPLUGIN_LOG_LEVEL",
+			Usage:  "set netplugin log level, options: [DEBUG, INFO, WARN, ERROR]",
+		},
+		cli.BoolFlag{
+			Name:   "use-json-log, json-log",
+			EnvVar: "CONTIV_NETPLUGIN_USE_JSON_LOG",
+			Usage:  "set netplugin log format to json",
+		},
+		cli.BoolFlag{
+			Name:   "use-syslog, syslog",
+			EnvVar: "CONTIV_NETPLUGIN_USE_SYSLOG",
+			Usage:  "set netplugin send log to syslog or not",
+		},
+		cli.StringFlag{
+			Name:   "syslog-url",
+			Value:  "udp://127.0.0.1:514",
+			EnvVar: "CONTIV_NETPLUGIN_SYSLOG_URL",
+			Usage:  "set netplugin syslog url in format protocol://ip:port",
+		},
+		cli.StringFlag{
+			Name:   "host, host-label",
+			Value:  hostname,
+			EnvVar: "CONTIV_NETPLUGIN_HOST",
+			Usage:  "set netplugin host to identify itself",
+		},
+		cli.StringFlag{
+			Name:   "mode, plugin-mode",
+			EnvVar: "CONTIV_NETPLUGIN_MODE",
+			Usage:  "set netplugin mode, options: [docker, kubernetes, swarm-mode]",
+		},
+		cli.StringFlag{
+			Name:   "vtep-ip",
+			Value:  localIP,
+			EnvVar: "CONTIV_NETPLUGIN_VTEP_IP",
+			Usage:  "set netplugin vtep ip for vxlan communication",
+		},
+		cli.StringFlag{
+			Name:   "ctrl-ip",
+			Value:  localIP,
+			EnvVar: "CONTIV_NETPLUGIN_CONTROL_IP",
+			Usage:  "set netplugin control ip for control plane communication",
+		},
+		cli.StringFlag{
+			Name:   "vlan-uplinks, vlan-if",
+			EnvVar: "CONTIV_NETPLUGIN_VLAN_UPLINKS",
+			Usage:  "a comma-delimited list of netplugin uplink interfaces",
+		},
+		cli.StringFlag{
+			Name:   "etcd-endpoints, etcd",
+			EnvVar: "CONTIV_NETPLUGIN_ETCD_ENDPOINTS",
+			Usage:  "a comma-delimited list of netplugin etcd endpoints",
+		},
+		cli.StringFlag{
+			Name:   "consul-endpoints, consul",
+			EnvVar: "CONTIV_NETPLUGIN_CONSUL_ENDPOINTS",
+			Usage:  "a comma-delimited list of netplugin consul endpoints",
+		},
+		cli.IntFlag{
+			Name:   "vxlan-port",
+			Value:  4789,
+			EnvVar: "CONTIV_NETPLUGIN_VXLAN_PORT",
+			Usage:  "set netplugin VXLAN port",
+		},
+		cli.StringFlag{
+			Name:   "netmode, network-mode",
+			EnvVar: "CONTIV_NETPLUGIN_NET_MODE",
+			Usage:  "set netplugin network mode, options: [vlan, vxlan]",
+		},
+		cli.StringFlag{
+			Name:   "fwdmode, forward-mode",
+			EnvVar: "CONTIV_NETPLUGIN_NET_MODE",
+			Usage:  "set netplugin forwarding network mode, options: [bridge, routing]",
+		},
+		/*
+			// only ovs is supported
+			cli.StringFlag {
+				Name: "driver, net-driver",
+				Value: "ovs",
+				EnvVar: "CONTIV_NETPLUGIN_DRIVER",
+				Usage: "set netplugin key-value store url, options: [ovs, vpp]",
+			}
+		*/
+	}
+	sort.Sort(cli.FlagsByName(app.Flags))
+	app.Action = func(ctx *cli.Context) error {
+		configs, err := initNetPluginConifg(ctx)
+		if err != nil {
+			errmsg := err.Error()
+			logrus.Error(errmsg)
+			return cli.NewExitError(errmsg, (len(errmsg)%254 + 1))
```

Comment:
Could you explain what is happening here?

Reply:
get error msg and print log, exit with error msg and error code.
There's trick that the return code is length of error msg mod 254 +1 to make sure it's an in 1-255 range

Reply:
Right, I understand what the code is doing, but why are we returning an error code based on the length of the string?  Doesn't seem very useful to me (there could be multiple error messages with the same length, etc.)

Reply:
because all errors returns the same code is a little bad, i wanted to refactor `core/error` to include cli & http return code but it's a little to much for this PR.

Reply:
We should subtree contivmodel and then we can fix the entire API in one PR 👍

Reply:
Refactoring `core/error` is one idea. Alternatively, I think `initNetPluginConifg` can return a unique error code for every error (using Go's `hash`) like
```
func initNetPluginConifg(ctx *cli.Context) (plugin.Config, error, uniqueErrorCode) {
```
instead of
```
func initNetPluginConifg(ctx *cli.Context) (plugin.Config, error) {
```

Reply:
the `core/error.go` needs to be refactored and will include a http return codes and cli return codes IMO, i'll put it later